### ISSUE TITLE
v2 discussion, 2024 edition!

### ISSUE THREAD

**Issue by saghul** [libuv/libuv#4622]
I've seen @vtjnash reference v2 a few times recently in issues / PRs, and I was wondering if the "v1 forever" discussion should be revisited.

The original v2 plans were overly ambitious (mea culpa, to a large extent) and not alienating the vast libuv user base felt like a compelling reason not to go for it.

For the sake of those not aware, libuv versioning is mostly semver, if we ignore dropping old platforms so it doesn't count towards "backwards compatibility". In addition, we do count changing the ABI as backwards compatibility.

Quickly skimming through the commits in master I see commits in mostly these categories:

- ABI cleanups
- API refinements (uv_file vs uv_os_fd_t, while should also be like 1, I'm putting it here for clarity)
- Internals unification (shared loop watchers for example)
- 2038 things?
- ???

IIRC @vtjnash had some Windows specific stuff around pipes (?), anything I'm missing?

Questions:

1) Is there any value in landing any subset of that as a v2? To me it feels like a "quality of life" v2 with the ABI cleanups (having forgotten the reserved fields in the loop structure was annoying) could be nice.

2) Do we have the time and energy for larger changes?

**Comment by saghul** [comment]
Ping @libuv/collaborators

**Comment by juanarbol** [comment]
I'm eager to help on anything you may need from my side.

**Comment by bnoordhuis** [comment]
My position hasn't really changed: v1.x works okay for most users, v2.x is not game-changing enough to merit an ecosystem rift.

I also don't feel like maintaining two release branches.

**Comment by vtjnash** [comment]
If we don't do v2 with the various ABI improvements already there, I will probably drop off as a maintainer of libuv at some point, since I don't care about maintaining the v1.x legacy support and moving our code back to v1.x would be a regression in features, API, and performance (as you noted above in the commits list). Eventually I also want to make some more correctness improvements to Windows which require breaking the ABI (or replacing the v1.x object internals with ones that just call malloc to get more space, which seems rather wasteful). But I also have interest in various performance improvements to multi-threaded cases, which nodejs most likely doesn't care about (aside from Electron), and that could also make for a compelling v2 for some users, or become a separate project.

**Comment by bnoordhuis** [comment]
I feel that but do you want to maintain v1.x and v2.x in parallel for the next 3 or 4 years<sup>1</sup>? Node can't switch until the current LTS goes EOL (April 2027) and I doubt they could upgrade to v2.x before the next LTS (April 2025, maintained until April 2028.)

I guess I could ask the TSC if they're willing to throw money at v1.x maintenance, otherwise I don't see any of us stepping up. Maintaining v0.10.x between 2014-2016 was a tedious slog; I'm not keen to do that again.

> But I also have interest in various performance improvements to multi-threaded cases, which nodejs most likely doesn't care about (aside from Electron), and that could also make for a compelling v2 for some users, or become a separate project.

That's something I'm thinking about from time to time. I don't want to derail the discussion here too much but maybe you and I can chat about that sometime.

<sup>1</sup> I appreciate that for you that is the current situation but in reverse

**Comment by santigimeno** [comment]
In general I'd like us to make our best to avoid having to split the project.

> I guess I could ask the TSC if they're willing to throw money at v1.x maintenance, otherwise I don't see any of us stepping up. Maintaining v0.10.x between 2014-2016 was a tedious slog; I'm not keen to do that again.

+1 on asking the TSC.

wrt the maintainance work, how different was `v0.10.x` from `v1.x`? I can't really remember, but what's proposed for `v2.x` doesn't seem too drastic of a change so that backporting stuff could be not too time consuming, but yeah I could be overly optimistic, and still the work would need to be done.

Some ideas to make the job not so bad:
-  Adding new maintainers to the team.
-  Not accepting new features in `v1.x`.

> 
> Loop me in!

+1

**Comment by vtjnash** [comment]
> I feel that but do you want to maintain v1.x and v2.x in parallel for the next 3 or 4 years?

That sounds a lot better than the current situation, where new features are expected to land constantly in v1.x and any improvements are expected to be backwards compatible in one branch but breaking in the other branch. I think the main question is whether we try to land some more breaking changes soon to give the API more room to grow in the future (e.g. make sure everything has flags or size arguments in the API)

> Node can't switch until the current LTS goes EOL (April 2027)

If they expect many years of support, it sounds like nodejs will need to pay the piper soon on gaining the 2038 support offered by the v2 upgrade (stat calls in v1 only support 32 bit time), or any nodejs programs running LTS are going to break in 15 years with no available path forward for them.

**Comment by bnoordhuis** [comment]
> Let's say we break the ABI but keep the API stable. Would it work then?

No. Libuv leaks out in node's public API/ABI.

> I could be overly optimistic

I think you'll come to hate it as time passes and the code bases diverge. I'll probably bow out at some point, even if the TSC begged me to take their money at or above my commercial rate. There are more satisfying things to work on.

> any nodejs programs running LTS are going to break in 15 years

I doubt 2038 is a pressing concern for node today. Maybe in another 5 years.

**Comment by bnoordhuis** [comment]
I've opened https://github.com/nodejs/TSC/issues/1657

**Comment by TheOneTheOnlyJJ** [comment]
Are there any updates on this?

**Comment by bnoordhuis** [comment]
This issue had its first anniversary two days ago. Do we still want to do this?

**Comment by saghul** [comment]
Dammit time goes fast.

I intended to grab a list of commits that are on master but not v1x so we could see what's in there.

**Comment by saghul** [comment]
Welcome to 2026 everyone :-)

So, seeing that @vtjnash continues to add cool stuff to master, I asked Claude to show us the current different between both branches, to better evaluate how a v2 would look like:

  | Commit | Description |
  |--------|-------------|
  | [`a00635af`](https://github.com/libuv/libuv/commit/a00635af) | Update version and SONAME for version 2 |
  | [`d010030a`](https://github.com/libuv/libuv/commit/d010030a) | Move all include files except `uv.h` to `uv/` |
  | [`68dafe90`](https://github.com/libuv/libuv/commit/68dafe90) | LEP-0005: remove `uv_file` type |
  | [`9b668ab2`](https://github.com/libuv/libuv/commit/9b668ab2) | Remove `uv_get_osfhandle` |
  | [`5ecd86c5`](https://github.com/libuv/libuv/commit/5ecd86c5) | Remove `uv_fs_poll_t` entirely (#5063) |
  | [`064617e4`](https://github.com/libuv/libuv/commit/064617e4) | Remove pipe shutdown logic, return `ENOTSUP` (#5067) |
  | [`c1775483`](https://github.com/libuv/libuv/commit/c1775483) | Expose loop reference in all `uv_req_t` |
  | [`1c91d2a6`](https://github.com/libuv/libuv/commit/1c91d2a6) | `uv_buf_init()` accepts `size_t` |
  | [`d7f65293`](https://github.com/libuv/libuv/commit/d7f65293) | Add reserved fields to `uv_loop_t` |
  | [`f8abdf67`](https://github.com/libuv/libuv/commit/f8abdf67) | Limit `uv_timer_again` to repeating timers |
  | [`9834c4f7`](https://github.com/libuv/libuv/commit/9834c4f7) | Return `UV_EINVAL` in `uv_##name##_start()` for invalid args |
  | [`daf8a45d`](https://github.com/libuv/libuv/commit/daf8a45d) | Return IOCP HANDLE in `uv_backend_fd` (Windows) |

  ## New Features (v2 only)

  | Commit | Description |
  |--------|-------------|
  | [`5047b344`](https://github.com/libuv/libuv/commit/5047b344) | Expose broadcast address on network interfaces (#2033) |
  | [`8eed38fd`](https://github.com/libuv/libuv/commit/8eed38fd) | Add `gecos` field to `uv_passwd_t` |
  | [`c51ead5c`](https://github.com/libuv/libuv/commit/c51ead5c) | Add `uv_fs_utime_ex` and `uv_fs_futime_ex` |
  | [`20cc969e`](https://github.com/libuv/libuv/commit/20cc969e) | Add CPU affinity mask option to `uv_spawn` |
  | [`6521405c`](https://github.com/libuv/libuv/commit/6521405c) | Add function to get CPU affinity mask size |
  | [`e5f79e48`](https://github.com/libuv/libuv/commit/e5f79e48) | Support setting supplementary process groups (#4704) |
  | [`8ad24655`](https://github.com/libuv/libuv/commit/8ad24655) | Allow users to configure child error mode (Windows) (#4569) |
  | [`da343e62`](https://github.com/libuv/libuv/commit/da343e62) | Allow fatal errors to trigger WER (Windows) |
  | [`00357f87`](https://github.com/libuv/libuv/commit/00357f87) | Don't run timers before entering event loop (#4026) |

  ## Internal Cleanup (v2 only)

  | Commit | Description |
  |--------|-------------|
  | [`912d42b7`](https://github.com/libuv/libuv/commit/912d42b7) | Unify timers implementation across unix/win |
  | [`3118e5a4`](https://github.com/libuv/libuv/commit/3118e5a4) | Share loop watchers implementation across unix/win |
  | [`3f21dd64`](https://github.com/libuv/libuv/commit/3f21dd64) | Remove ABI compatibility hacks (Windows) |
  | [`c82eedd0`](https://github.com/libuv/libuv/commit/c82eedd0) | Remove 1.x-compat code from fallback barrier |
  | [`e1d6e62a`](https://github.com/libuv/libuv/commit/e1d6e62a) | Track each `write_req` in a queue instead of count (Windows) |

**Comment by vtjnash** [comment]
It forgot to mention the 2038 bug fixes. Also notable (off the top of my head) that many errno are corrected on Windows to match Unix.

**Comment by saghul** [comment]
@bnoordhuis Would you be open to discussing v2, iff Node (plus maybe some members from other communities relying on it) takes care of maintaining v1.x until they EOL the last Node version using it?

I think that can work.

**Comment by bnoordhuis** [comment]
I've moved past caring for node so sure, go ahead.

**Comment by saghul** [comment]
Let's do this!

**Comment by santigimeno** [comment]
First of all, I'm personally in favour of doing a v2.
Now wearing my Node.js collaborator and NodeSource employee hats, I'm happy to help maintain the v1.x line.  For this I think we should define some kind of rough roadmap for v2 as it would help us decide:
- Which Node.js major could realistically ship libuv v2 for the first time.
- How long we should expect v1.x maintenance to continue.

I don't think we're in time to get into v26.x as it's going to be cut next week but v27.x should be a realistic goal.

> iff Node (plus maybe some members from other communities relying on it) takes care of maintaining v1.x until they EOL the last Node version using it

In your opinion, is this currently a blocker for v2 to happen?

Also, there's a Node.js Collaboration Summit in London next month I'm attending, so I can bring this topic there and get feedback from other Node.js maintainers.

**Comment by saghul** [comment]
> Now wearing my Node.js collaborator and NodeSource employee hats, I'm happy to help maintain the v1.x line.

Sounds great!

> * Which Node.js major could realistically ship libuv v2 for the first time.

I'd keep this open until we make a 2.0.0 release, there is no need to rush is there?

> * How long we should expect v1.x maintenance to continue.

Ideally until people using it in active LTS projects stop doing so. Node is a big one, but there are others too.

> > iff Node (plus maybe some members from other communities relying on it) takes care of maintaining v1.x until they EOL the last Node version using it
> 
> In your opinion, is this currently a blocker for v2 to happen?

Not for me!

> Also, there's a Node.js Collaboration Summit in London next month I'm attending, so I can bring this topic there and get feedback from other Node.js maintainers.

Sounds like a good idea!

**Comment by jasnell** [comment]
I'm also happy to help with maintenance on the v1.x line up through whenever the last supported Node.js LTS line ships it; and happy to do what I can to help get v2.x into Node.js.

Is there a list of TODO's for What's Left To Do For v2.x?

**Comment by saghul** [comment]
Thanks for chiming in @jasnell !



> Is there a list of TODO's for What's Left To Do For v2.x?

We should start collecting a list. Maybe this is the right place. I think at the moment it's all spread in our collective minds 😅

**Comment by vtjnash** [comment]
Theoretically we should do most of these: https://github.com/libuv/libuv/issues?q=is%3Aissue%20state%3Aopen%20label%3Av2. Though we don't need to actually block v2 on having completed all of them.

We might also want to decide on how to address timer mis-firing (https://github.com/libuv/libuv/issues/4773)

**Comment by TheOneTheOnlyJJ** [comment]
@vtjnash I was looking around the ongoing work here and have a few curiosities about this comment you made in 2024:

> Eventually I also want to make some more correctness improvements to Windows which require breaking the ABI (or replacing the v1.x object internals with ones that just call malloc to get more space, which seems rather wasteful).

What improvements are you referring to? What is the status of these? I see you did a lot of Windows-related work since then. Is it related to this comment?

> But I also have interest in various performance improvements to multi-threaded cases, which nodejs most likely doesn't care about (aside from Electron), and that could also make for a compelling v2 for some users, or become a separate project.

Again, I am curious about what you are referring to here. Especially about the Electron mention. What could be done here that is of interest to Electron? Also curious about the status of this idea.

**Comment by vtjnash** [comment]
Yes, #5059 (make write cancellation handled correctly) and #5078 (make 64 bit handled correctly) were the main ones. There is also https://github.com/libuv/libuv/issues/2016, where any process with a copy of any libuv handle can be used to cause libuv to perform arbitrary memory access (and the proposed fix for it requires #5059)

Electron is partly multi-threaded, on Windows specifically, because it needs to run 2 event loops, and Window's kernel can't do that (IOCP is not a waitable handle type). So it does a complicated little song and dance on every loop iteration to run the loop on a different thread, then copy the events back to the first thread. Or at least, this was true 10 years ago since I've had no reason to look again since about 5 years ago.

**Comment by santigimeno** [comment]
I was looking through https://github.com/libuv/leps/pull/2. Do we want to pursue this (either fully or partially)?

**Comment by saghul** [comment]
> I was looking through [libuv/leps#2](https://github.com/libuv/leps/pull/2). Do we want to pursue this (either fully or partially)?

IMHO it's too big a change it it would create quite some churn.

(my) initial idea to re-open the v2 discussion was to do mostly ABI cleanup, not a huge change.

**Comment by santigimeno** [comment]
> IMHO it's too big a change it it would create quite some churn.

Don't disagree though some parts of it don't seem so invasive: `uv_timeout` or `uv_callback`.

> (my) initial idea to re-open the v2 discussion was to do mostly ABI cleanup, not a huge change.

Got it. So maybe for v3 in 2035 😆

**Comment by saghul** [comment]
> Don't disagree though some parts of it don't seem so invasive: `uv_timeout` or `uv_callback`.

Right, I had uv_read in mind when I read your comment. Sure, I think we could consider those!

> Got it. So maybe for v3 in 2035 😆

Are we unofficially comitting to a "one major every 10 years" ? :-P

**Comment by santigimeno** [comment]
> Are we unofficially committing to a "one major every 10 years" ? :-P

My bad! That wasn't my intention at all. Maybe a middle ground can be found: what about opening "v3 discussion" in 2036? It should give us enough leeway.

**Comment by santigimeno** [comment]
> > Don't disagree though some parts of it don't seem so invasive: `uv_timeout` or `uv_callback`.
> 
> Right, I had uv_read in mind when I read your comment. Sure, I think we could consider those!

Ok. I'll work on `uv_timeout`.

**Comment by saghul** [comment]
> > Are we unofficially committing to a "one major every 10 years" ? :-P
> 
> My bad! That wasn't my intention at all. Maybe a middle ground can be found: what about opening "v3 discussion" in 2036? It should give us enough leeway.
> 
> 

I was joking! 🤣

**Comment by santigimeno** [comment]
> I was joking! 🤣

I know! So was I! (forgot the emoji)
